[explore] Fix and test slice id  logging  issue

timifasubaa opened the pull request:
This PR fixes and tests the slice_id logging issue (https://github.com/apache/incubator-superset/issues/3311) 
@mistercrunch @john-bodley

I was able to successfully test the /explore case but I had issues with the ajax call (or /explore_json) case. The error was CSRF token is missing. I might be missing something either in the login or in the process of making the request for JSON. The login call triggers the error. This is the same way other functions that needed to login did so.

If you want to dig in, the stack trace I get is : 
```
Traceback (most recent call last):
  File "/Users/timi_fasubaa/Desktop/incubator-superset/tests/core_tests.py", line 739, in test_slice_id_is_always_logged_correctly_on_ajax_request
    self.login(username="admin")
  File "/Users/timi_fasubaa/Desktop/incubator-superset/tests/base_tests.py", line 131, in login
    self.assertIn('Welcome', resp)
AssertionError: u'Welcome' not found in u'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>400 Bad Request</title>\n<h1>Bad Request</h1>\n<p>The CSRF token is missing.</p>\n'
    """Fail immediately, with the given message."""
>>  raise self.failureException('u\'Welcome\' not found in u\'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\\n<title>400 Bad Request</title>\\n<h1>Bad Request</h1>\\n<p>The CSRF token is missing.</p>\\n\'')
```

timifasubaa committed:
[explore] fixed padding bug on filter section

timifasubaa committed:
fix slice_id logging issue

timifasubaa committed:
[superset-sqllab] fix slice_id population in appropriate column

timifasubaa committed:
[explore-logging] test the slice_id logging fix

timifasubaa committed:
Merge branch 'master' of https://github.com/apache/incubator-superset

mistercrunch commented:
see how the other tests include a `self.login` call.

mistercrunch commented:
Also looking at Travis build logs it looks like json.loads doesn't like `None`

timifasubaa commented:
I took inspiration from test_user_profile and test_fetch_datasource_metadata and what I have as far as I can see is consistent with what they do. 
Strangely, when I run 
```
nosetests tests.core_tests:CoreTests.test_user_profile
nosetests tests.core_tests:CoreTests.test_fetch_datasource_metadata
```
They both  error out with the same CSRF issue. Assuming both tests worked before, the fault may lie somewhere in the underlying authentication logic on my computer and not the tests. Can you please try on yours and see if the issue is still there?

mistercrunch commented:
Why nosetests over `./run_specific_test.sh`? It sets env vars you need.

xrmx commented on superset/models/core.py:
```diff
@@ -761,8 +761,8 @@ def wrapper(*args, **kwargs):
             d.update(kwargs)
-            slice_id = d.get('slice_id', 0)
             try:
+                slice_id = d.get('slice_id') or json.loads(d.get('form_data')).get('slice_id')
```
This will not catch when you don't have nor slice_id neither form_data:
https://travis-ci.org/apache/incubator-superset/jobs/266166713#L4907

timifasubaa committed:
fix travis errors

timifasubaa commented:
@mistercrunch  I don't know why nosetests came to mind when I was on the terminal. I tried out ./run_specific_test.sh and it worked fine.

@xrmx  I also addressed the possible issue with the case where neither is filled in. I don't know if making slice_id 0 after the exception is thrown is the best solution. Anything that doesn't log the right slice_id will fail the tests I wrote. 

On line 73 in core_tests.py, I called explunge_all() to store things out of a session. I don't understand why things were out of the session in the first place (all other similar calls to get_resp() never had to call that function but after I did, it worked. Am I doing something wrong there?

john-bodley commented:
Personally I would have kept what you had but wrapped in a int(...) call per https://github.com/apache/incubator-superset/issues/3311#issuecomment-322939564.

timifasubaa commented:
@john-bodley I was in agreement until @xrmx 's comment above. The reason I changed it is that in the event there is not form_data and json.loads takes in None, it will throw a TypeError, which won't be caught by the existing catch block. Would you prefer I changed it to also catch ValueError in addition to Exception?

mistercrunch commented on tests/core_tests.py:
```diff
@@ -726,7 +727,23 @@ def test_user_profile(self, username='admin'):
+        slc = db.session.query(models.Slice).filter_by(slice_name='Girls').one()
+        db.session.expunge_all()
+        self.get_resp(slc.slice_url)
+        qry = db.session.query(models.Log).filter_by(slice_id=slc.id, )
```
typically no trailing comma on function call

mistercrunch commented on superset/models/core.py:
```diff
@@ -761,8 +761,11 @@ def wrapper(*args, **kwargs):
             d.update(kwargs)
-            slice_id = d.get('slice_id', 0)
             try:
+                slice_id = d.get('slice_id') 
```
it's better to keep try blocks as nimble as possible, 765:766 are safe (`dict.get` won't throw key errors if the key doesn't exist it'll just return `None`) and can be executed outside the try block.

mistercrunch commented on tests/core_tests.py:
```diff
@@ -726,7 +727,23 @@ def test_user_profile(self, username='admin'):
+        # superset/explore_json case
+        self.login(username="admin")
+        slc = db.session.query(models.Slice).filter_by(slice_name='Girls').one()
+        qry = db.session.query(models.Log).filter_by(slice_id=slc.id, )
```
same here, no trailing coma on method calls

mistercrunch replied:
NIT: I'd move this to right above the `self.assertEqual(1, qry.count())` line as the variable doesn't need to exist while the request takes place. Shorter lived variables are better in general. It's pretty minor but could cause confusion as to when the query is actually executed.

mistercrunch commented on superset/models/core.py:
```diff
@@ -761,8 +761,11 @@ def wrapper(*args, **kwargs):
+                if not slice_id and form_data:
+                    slice_id = json.loads(form_data).get('slice_id')                    
                 slice_id = int(slice_id) if slice_id else 0
             except ValueError:
```
json.loads might raise something else than `ValueError`, for instance `json.loads(None)` raises `TypeError`. You might just want to be safe and catch all exceptions here as we're not trying to handle a specific type of error but just keep the program flowing regardless of what error may be raised.

mistercrunch commented:
Also rebasing may help with coveralls checks

timifasubaa committed:
fix nits pointed out in PR comments

timifasubaa committed:
merge from upstream

timifasubaa committed:
cleanup tests

mistercrunch commented:
LGTM

xrmx commented on superset/models/core.py:
```diff
@@ -761,24 +761,26 @@ def wrapper(*args, **kwargs):
             try:
-                slice_id = int(slice_id) if slice_id else 0
-            except ValueError:
+                slice_id = int(d.get('slice_id') or json.loads(d.get('form_data')).get('slice_id'))
```
Isn't this ugly? You have slice_id and form_data already unpacked before and you are not using them here. That would make the code a lot more readable.

Also blind exceptions are an antipattern. If you want to go this route just catch TypeError and ValueError

timifasubaa replied:
Would yo prefer that I break it out in the if else style I had before that doesn't use exception for flow control? I am trying out the ValueError and TypeError approach now.

timifasubaa committed:
made python more beautiful

xrmx commented on superset/models/core.py:
```diff
@@ -761,24 +761,26 @@ def wrapper(*args, **kwargs):
-                slice_id = int(slice_id) if slice_id else 0
-            except ValueError:
+                slice_id = int(slice_id or json.loads(form_data).get('slice_id'))
+            except Exception:
```
Please remove blind exceptions, You only need ValueError and TypeError

timifasubaa committed:
made python even more beautiful

xrmx reviewed (commented):
It's not, for sure not pep8. Also you fat fingered www

timifasubaa committed:
made python even more more beautiful

timifasubaa committed:
made python even more more more beautiful

timifasubaa commented:
@xrmx  You're right. I reviewed and updated it. What do you think of the current state?

timifasubaa committed:
fix lint error

xrmx commented:
@timifasubaa i think that you should fix issues pointed by the reviewer instead

timifasubaa committed:
make exception handling more specific

timifasubaa commented:
@xrmx how about now?

timifasubaa committed:
fixed silly error

timifasubaa committed:
fixed argument indentation

mistercrunch commented:
LGTM, merging
